pkg/targets/adapter/mongodbtarget/adapter.go

```diff
@@ -0,0 +1,250 @@
+	sr       *metrics.EventProcessingStatsReporter
+}
+
+func (a *adapter) Start(ctx context.Context) error {
+	a.logger.Info("Starting MongoDB adapter")
+	return a.ceClient.StartReceiver(ctx, a.dispatch)
+}
+
+func (a *adapter) dispatch(event cloudevents.Event) (*cloudevents.Event, cloudevents.Result) {
+	ctx := context.Background()
+	ceTypeTag := metrics.TagEventType(event.Type())
+	ceSrcTag := metrics.TagEventSource(event.Source())
+	a.logger.Debug("Processing event:" + event.Type())
+	start := time.Now()
+	defer func() {
+		a.sr.ReportProcessingLatency(time.Since(start), ceTypeTag, ceSrcTag)
+	}()
+
+	switch typ := event.Type(); typ {
+	case v1alpha1.EventTypeMongoDBInsert:
+		if err := a.insert(event, ctx); err != nil {
+			a.logger.Errorw("invoking .insert: ", zap.Error(err))
+			a.sr.ReportProcessingError(true, ceTypeTag, ceSrcTag)
+			return a.replier.Error(&event, targetce.ErrorCodeAdapterProcess, err, nil)
+		}
+	case v1alpha1.EventTypeMongoDBQueryKV:
+		resp, err := a.kvQuery(event, ctx)
+		if err != nil {
+			a.logger.Errorw("invoking .query.kv: ", zap.Error(err))
+			a.sr.ReportProcessingError(true, ceTypeTag, ceSrcTag)
+			return a.replier.Error(&event, targetce.ErrorCodeAdapterProcess, err, nil)
+		}
+		return resp, nil
+	case v1alpha1.EventTypeMongoDBUpdate:
+		if err := a.update(event, ctx); err != nil {
+			a.logger.Errorw("invoking .update: ", zap.Error(err))
+			a.sr.ReportProcessingError(true, ceTypeTag, ceSrcTag)
+			return a.replier.Error(&event, targetce.ErrorCodeAdapterProcess, err, nil)
+		}
+	default:
+		if err := a.insertFromConfig(event, ctx); err != nil {
+			a.logger.Errorw("invoking arbirary insert: ", zap.Error(err))
+			a.sr.ReportProcessingError(true, ceTypeTag, ceSrcTag)
+			return a.replier.Error(&event, targetce.ErrorCodeAdapterProcess, err, nil)
+		}
+	}
+	return a.replier.Ok(&event, "ok")
+}
+
+// insertFromConfig inserts a mongodb document using the provided collection and database
+// at the spec configuration.
+func (a *adapter) insertFromConfig(e cloudevents.Event, ctx context.Context) error {
+	var payload map[string]interface{}
+	if err := e.DataAs(&payload); err != nil {
+		return err
+	}
+
+	collection := a.mclient.Database(a.defaultDatabase).Collection(a.defaultCollection)
+	_, err := collection.InsertOne(ctx, payload)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// kvQuery queries a mongodb collection for a specific key/value pair.
+func (a *adapter) kvQuery(e cloudevents.Event, ctx context.Context) (*cloudevents.Event, error) {
+	qpd := &QueryPayload{}
+	if err := e.DataAs(qpd); err != nil {
+		return nil, err
+	}
+	col := a.defaultCollection
+	db := a.defaultDatabase
+	if qpd.Collection != "" {
+		col = qpd.Collection
+	}
+	if qpd.Database != "" {
+		db = qpd.Database
+	}
+
+	collection := a.mclient.Database(db).Collection(col)
+	filterCursor, err := collection.Find(ctx, bson.M{qpd.Key: qpd.Value})
+	if err != nil {
+		return nil, err
+	}
+
+	var episodesFiltered []bson.M
```

Comment:
Is episodes some sort of new mongo lang?

Reply:
no. this should be revised.